R/backends.R

```diff
@@ -957,3 +944,56 @@ read_csv_as_stanfit <- function(files, variables = NULL,
     .MISC = new.env(parent = emptyenv())
   )
 }
+
+
+#' Read CmdStan CSV files as a brms-formatted stanfit object
+#'
+#' `read_csv_as_brms_stanfit()` is used internally to read CmdStan CSV files into a
+#' `stanfit` object that is consistent with the structure of the fit slot of a
+#' brmsfit object.
+#'
+#' @param files character vector of CSV files names where draws are stored
+#' @param model a compiled cmdstanr model object (optional). Provide this argument
+#'  if you want to allow updating the model without recompilation.
+#' @param exclude character vector of variables to exclude from the stanfit
+#'
+#' @return a stanfit object consistent with the structure of the fit slot of
+#'  a brmsfit object
+#' @export
+#'
+#' @examples
+#' \dontrun{
+#' # fit a model manually via cmdstanr
+#' scode <- stancode(count ~ Trt, data = epilepsy)
+#' sdata <- standata(count ~ Trt, data = epilepsy)
+#' mod <- cmdstanr::cmdstan_model(cmdstanr::write_stan_file(scode))
+#' stanfit <- mod$sample(data = sdata)
+#'
+#' # feed the Stan model back into brms
+#' fit <- brm(count ~ Trt, data = epilepsy, empty = TRUE, backend = 'cmdstanr')
+#' fit$fit <- read_csv_as_brms_stanfit(stanfit$output_files(), model = mod)
+#' fit <- rename_pars(fit)
+#' summary(fit)
+#' }
+read_csv_as_brms_stanfit <- function(files, model, exclude = "") {
+  # not all metadata is stored by read_csv_as_stanfit
+  require_package("cmdstanr")
+  metadata <- cmdstanr::read_cmdstan_csv(
```

Comment:
reading the metadata manually like this introduces unnecessary overhead. this information is available in only a slightly different format from `out$metadata()`.  This information is used to select the variables, which are then renamed and combined with exclude, before being passed to read_csv_as_stanfit. Which are then renamed in the same way in[ that function as well.](https://github.com/paul-buerkner/brms/blob/7b0adcd75e68bef8546c439880ba606bed8cb29b/R/backends.R#L659), so this is unnecessary.

I think the whole thing can be simplified into a single function with less duplication of code and less unnecessary reading of files. I may be missing why this is done this way, so I'll make another version in a separate branch and you can decide which one to keep

Reply:
Here's the alternative version: https://github.com/paul-buerkner/brms/pull/1619